### ISSUE TITLE
Ion and electron transport iterative problem

### ISSUE THREAD

**Issue by Proovesor** [lululxvi/deepxde#336]
Hello Dr Lu.

I've been recently using your library in my research, which mainly concerns solving ionic and electronic transport in the Solid Oxide Fuel Cell anode. More specifically, I'm trying to solve equations (7) and (8), based on Butler-Volmer equation (19) ([article ref](https://www.sciencedirect.com/science/article/pii/S0378775311000668)).

Ion and electron transport:

<img src="https://render.githubusercontent.com/render/math?math=\nabla\cdot(\sigma^{\mathrm{eff}}_{\mathrm{el}}\nabla\phi_\mathrm{el})=i_\mathrm{tpb}"> 
<img src="https://render.githubusercontent.com/render/math?math=\nabla\cdot(\sigma^{\mathrm{eff}}_{\mathrm{io}}\nabla\phi_\mathrm{io})=-i_\mathrm{tpb}"> 

Butler Volmer:

<img src="https://render.githubusercontent.com/render/math?math=i_\mathrm{tpb}=i_0\left[\exp{\frac{\beta nF}{RT}\eta_\mathrm{act}}-\exp{-\frac{(1-\beta)nF}{RT}\eta_\mathrm{act}}\right]">

The main question is, is iterative way of model learning possible? Meaning that based on derived potentials from (7) and (8) I then calculate value of i_tpb (Butler Volmer), then the model runs again, with new values of potentials. The desired output should look like fig. 7 in the attached article.

I've done many approaches, but in order to make things easier, I've assumed phi_el = 0,05V (later on I'd like to simultaneously calculate phi_el as one component and phi_io as second component).
The main struggles are that I can't force boundary conditions to actually stay at given value and am totally not sure of the right way of calculating Butler Volmer and feeding PDE with it. I might have totally missed the point or can't quite grasp the idea and concept of how could and should this be working.

I've managed to produce such output with code below (which is one of many apporaches), which I'd be thrilled to discuss of how to make it right, as the results haven't been satisfying.

![issue_png](https://user-images.githubusercontent.com/11725909/125806941-cff53c5a-2eae-4371-8abf-3203e0bcccac.png)

While this might look promising, the scale isn't actually giving best look at it, so here is array of derived potentials:
```
[0.02500069 0.02500069 0.02500068 0.02500062 0.02500068 0.02500072 0.0250006  0.02500063 0.02500066 0.02500066 0.02500071 0.02500063 0.02500068 0.02500059 0.02500062 0.02500062 0.02500063 0.02500056 0.02500057 0.02500053 0.02500059 0.02500053 0.02500059 0.0250005 0.0250005  0.02500053 0.02500051 0.02500047 0.02500042 0.02500036 0.02500039 0.0250003  0.02500029 0.02500029 0.02500026 0.02500018 0.02500013 0.02500008 0.02500007 0.02500002 0.02500002 0.02499993 0.02499986 0.02499983 0.02499975 0.02499966 0.0249996  0.02499953 0.02499953 0.02499938]
```

which shows that potential clearily stays in between of imposed DirichletBCs.

</body>
</html>

```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import deepxde as dde

from deepxde.boundary_conditions import DirichletBC
from deepxde.boundary_conditions import NeumannBC
from deepxde.geometry.geometry_1d import Interval
from deepxde.data.pde import PDE
from deepxde.maps.fnn import FNN
from deepxde.model import Model
from deepxde.postprocessing import saveplot

import numpy as np
from matplotlib import pyplot as plt
from deepxde.backend import tf

# PARAMS
p_t = 1.013e5 # Pa
temperature = 1073 # K
p_H2 = 0.97 * p_t # Pa
p_H2O = 0.03 * p_t # Pa
Phi_el = 0.05 # V
sigma_el = 1.43e7 # S m-1
sigma_io = 18 # S m-1
beta_const = 0.5
n = 2
gas_const = 8.314462 # J K-1 mol-1
faraday_const = 96485.33212 # s A mol-1
tau_io = 8.85
V_io = 0.251
tau_el = 6.91
V_el = 0.253

# DIMS
X = 5e-5 # m
l_tpb = 2.49e12 # m m-3

num_domain = 10
num_test = 50
num_boundary = 4

# empirical formula for exchange current per tpb length
def i_exchanged():
    i0_tpb = 31.4 * p_H2**-0.03 * p_H2O**0.4 * np.exp(-18300 / temperature)
    return i0_tpb * l_tpb

# butler-volmer equation
def butler_volmer(io_potential):
    i_0 = i_exchanged()
    activation_overpotential = (np.ones(num_test) * Phi_el - io_potential)

    return [i_0 * (np.exp(beta_const * n * faraday_const * act / (gas_const * temperature)) - np.exp(-(1 - beta_const) * n * faraday_const * act / (gas_const * temperature))) for act in activation_overpotential]

def main(potentials_input=[]):

    def pde(x, y):
        sigma_eff_io = V_io / tau_io * sigma_io
        dphi_io_x = tf.gradients(y, x)[0]
        dphi_io_xx = tf.gradients(dphi_io_x, x)[0]

        io_potentials = np.linspace(0.05, 0, num=num_test)
        if len(potentials_input) > 0:
            io_potentials = potentials_input

        io_equation = dphi_io_xx * sigma_eff_io + butler_volmer(io_potentials) * x

        return io_equation

    def left_boundary(x, on_boundary):
        return np.isclose(x, 0) and on_boundary

    def right_boundary(x, _):
        return np.isclose(x, X)

    geometry = Interval(0, X)

    left_io_neu = dde.NeumannBC(geometry, lambda x: np.full_like(x, 0), left_boundary)
    right_io = dde.DirichletBC(geometry, lambda x: np.full_like(x, 0), right_boundary)
    left_io_dir = dde.DirichletBC(geometry, lambda x: np.full_like(x, 0.05), left_boundary)

    layers = [1] + [40] * 3 + [1]
    activation = 'tanh'
    initializer = 'Glorot uniform'
    data = PDE(geometry, pde, [left_io_neu, right_io, left_io_dir], num_domain=num_domain, num_boundary=num_boundary, num_test=num_test)
    net = FNN(layers, activation, initializer)

    model = Model(data, net)

    model.compile("adam", lr=0.001, loss_weights=[1e-8, 1e-2, 1, 1])
    losshistory, train_state = model.train(epochs=200000)

    return [losshistory, train_state]

if __name__ == '__main__':
    predicted_potentials = []
    iter = 1

    for i in range(iter):
        [losshistory, train_state] = main(np.array(predicted_potentials).flatten())
        X_train, y_train, X_test, y_test, best_y, best_ystd = train_state.packed_data()
        predicted_potentials = best_y

        # if i == iter - 1:
        #     saveplot(losshistory, train_state, issave=True, isplot=True)

```

I'd be really happy to hear any suggestions as, as I've stated above, I might've completely lost some concepts.

**Comment by lululxvi** [comment]
The iterative way could be doable, but I am not sure and I haven't done this before. What is the motivation of solving it iteratively? Why don't you just plug the i_tpb into the PDE?

Also, your parameters have different scales, and it is always to scale them to be O(1).

Another issue I find is `left_boundary` and `right_boundary`. `x` is a d-dim vector, so
```
    def left_boundary(x, on_boundary):
        return np.isclose(x[0], 0) and on_boundary
```

**Comment by Proovesor** [comment]
Thanks for you reply! And sorry for late respond, I was out of town.

The motivation is so, because the potentials are obtained based on butler volmer equation and since equations (7) and (8) include i_tpb, I thought iteratively is the way to do this. Right now i_tpb is being plugged into PDE with some initial values set, but I can't obtain any reasonable results.

Sorry, but I can't quite understand at which point are the scales wrong, could be more specific? :)

Fixed boundary conditions definition, but nothing particular changed.

**Comment by lululxvi** [comment]
For example,  p_t = 1.013e5. The network is very difficult to learn things like this. In deep learning, it is always required to normalize the problem, e.g., https://machinelearningmastery.com/how-to-improve-neural-network-stability-and-modeling-performance-with-data-scaling/ (this link is data fitting, but the idea is the same). You can also check FAQ.

**Comment by Proovesor** [comment]
Thank you very much. Will give it a try!
